securedrop/journalist.py

```diff
@@ -333,6 +349,162 @@ def edit_account():
+
+
+@app.route('/admin/delete_source_label_type/<int:tag_id>', methods=('POST',))
+@admin_required
+def admin_delete_source_label_type(tag_id):
+    delete_source_label(tag_id)
+    return redirect(url_for('admin_index'))
+
+
+@app.route('/admin/create_source_label_type/', methods=('POST',))
+@admin_required
+def admin_create_source_label_type():
+    text_to_display = request.form['text_to_display']
+    all_labels = [x.label_text for x in get_all_defined_source_tags()]
+    if text_to_display not in all_labels:
+        db_session.add(SourceLabelType(label_text=text_to_display))
+        db_session.commit()
+    else:
+        flash('Tag already exists!', 'source')
+    return redirect(url_for('admin_index'))
+
+
+@app.route('/admin/create_submission_label_type/', methods=('POST',))
+@admin_required
+def admin_create_submission_label_type():
+    text_to_display = request.form['text_to_display']
+    all_labels = [x.label_text for x in get_all_defined_submission_tags()]
+    if text_to_display not in all_labels:
+        db_session.add(SubmissionLabelType(label_text=text_to_display))
+        db_session.commit()
+    else:
+        flash('Tag already exists!', 'submission')
+    return redirect(url_for('admin_index'))
+
+
+@app.route('/admin/delete_submission_label_type/<int:tag_id>', methods=('POST',))
+@admin_required
+def admin_delete_submission_label_type(tag_id):
+    delete_submission_label(tag_id)
+    return redirect(url_for('admin_index'))
+
+
+def create_source_tag(source_id, label_id):
+    source = Source.query.filter(Source.id == source_id).first()
+    db_session.add(SourceTag(source_id=source.id,
+                             label_id=label_id))
+    db_session.commit()
+
+
+def delete_source_tag(source_id, label_id):
+    source = Source.query.filter(Source.id == source_id).first()
+    matching_tag = SourceTag.query.filter(and_(SourceTag.label_id == label_id,
+                                               SourceTag.source_id == source.id)).all()
+    for tag in matching_tag:
+        db_session.delete(tag)
+    db_session.commit()
+
+
+def create_submission_tag(submission_id, label_id):
+    submission = Submission.query.filter(Submission.id == submission_id).first()
+    db_session.add(SubmissionTag(submission_id=submission.id,
+                                 label_id=label_id))
+    db_session.commit()
+
+
+def delete_submission_tag(submission_id, label_id):
+    submission = Submission.query.filter(Submission.id == submission_id).first()
+    matching_tag = SubmissionTag.query.filter(and_(SubmissionTag.label_id == label_id,
+                                                   SubmissionTag.submission_id == submission.id)).all()
+    for tag in matching_tag:
+        db_session.delete(tag)
+    db_session.commit()
+
+
+@app.route("/col/add_source_label/<sid>/<label_id>", methods=('POST',))
+@login_required
+def add_source_label(sid, label_id):
+    source = get_source(sid)
+    create_source_tag(source.id, label_id)
```

Comment:
In general, since we are using an ORM, it is preferable to pass around objects returned from the ORM (representing rows in the database) instead of primary keys, like you are doing here.